### DISCUSSION TITLE
How to adopt PID factor for the DriveBase?

### DISCUSSION THREAD

**Discussion by JohnGottiGe** [pybricks/support#1418] [General]
Hi Everybody,

I'm new with pybricks and did some test wit the DriveBase Class, setting use_gyro(TRUE). Additional our brick is build in with an other direction (top_side=Axis.Z, front_side=Axis.Y).
When now using DriveBase.straight the robo is not driving straight, it is curving and drifting to one side. I searched the documentation to find something to adopt the PID Factor, the only thing I found was the adoption for the motor PID factor using control.pid.
Is there something possible for the DriveBase? Or is there anything to adopt gyro to the robo? How can I influence the drive so that the robo will drive straight without a drift?
Thanks for your support.

**Comment by laurensvalk** [comment]
Can you share a video of it drifting to one side? Is it a lot?

Make sure to leave the robot sitting still for a few seconds at least once after you turn the robot on.

**Reply by JohnGottiGe** [reply]
I'm using this short program to drive straight, make a turn, drive back and turn again. Theoretical I should reach nearly the same position.
```python
def p_5_run():
    print("YELLOW = P5")
    # myHub.speaker.beep(300, 300)
    robo.straight(400, then=Stop.COAST, wait=True)
    print(f'Hinfahrt = {robo.distance()} Winkel = {myHub.imu.heading()}')
    robo.turn(180, wait=True)
    wait(500)
    robo.straight(400, wait=True)
    wait(500)
    print(f'Rückfahrt = {robo.distance()} Winkel = {myHub.imu.heading()}')
    robo.turn(180, then=Stop.COAST, wait=True)
    print(f'Winkel = {myHub.imu.heading()}')
    # robo.curve(400, -90, then=Stop.BRAKE, wait=True)
    myHub.speaker.beep(500, 100)
    return
```
The console output is also shown in the attached [screenshot.]([url](

https://github.com/pybricks/support/assets/89180965/6d33399f-c73b-4ff9-8c4e-5d7d8f647640

![output](https://github.com/pybricks/support/assets/89180965/ffbb39d9-675e-43fb-8a1d-3e663c3d6548)))
Attached the video

**Reply by laurensvalk** [reply]
The light seems to indicate that the batteries are nearly empty, so that would be the first thing to look at.

That initial shake while driving doesn't look quite right. The center of gravity also does appear quite far back which means the motors have to work harder, all the while having less grip. Also be sure to check the friction at the support points. Also, did you configure the axle track correctly? A gyro can help, but fixing the mechanics is always preferred if possible :)

**Reply by JohnGottiGe** [reply]
Hi Laurens,
thanks a lot for your feedback. Battery is okay, the light is just because we use a light sensor to detect attachments color to select FLL runs automatically by color. In this case I just used yellow "attachment" to do my tests.
Points regarding center of gravity and resulting friction are good points which I will check.
Thanks
